backend/elasticsearch-client/common/src/main/java/org/hibernate/search/backend/elasticsearch/client/common/spi/ElasticsearchRequestInterceptorContext.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright Red Hat Inc. and Hibernate Authors
+ */
+package org.hibernate.search.backend.elasticsearch.client.common.spi;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.List;
+import java.util.Map;
+
+import org.hibernate.search.util.common.annotation.Incubating;
+
+@Incubating
+public interface ElasticsearchRequestInterceptorContext {
+
+	boolean hasContent();
+
+	InputStream content() throws IOException;
+
+	String scheme();
+
+	String host();
+
+	Integer port();
+
+	String method();
+
+	String path();
+
+	Map<String, String> queryParameters();
+
+	void overrideHeaders(Map<String, List<String>> headers);
```

Comment:
for aws signing I just reshuffled the existing code to remove any apache specifics and push them to the client impls. Since we just need to set a few headers that signing might produce, this seems pretty straightforward.